build-system/compile/compile.js

```diff
@@ -167,249 +225,232 @@ function compile(
-    }
-
-    const compilerOptionsArray = [];
-    Object.keys(compilerOptions).forEach(function (option) {
-      const value = compilerOptions[option];
-      if (value instanceof Array) {
-        value.forEach(function (item) {
-          compilerOptionsArray.push('--' + option + '=' + item);
-        });
+      if (value != null) {
+        compilerFlags.push('--' + option + '=' + value);
       } else {
-        if (value != null) {
-          compilerOptionsArray.push('--' + option + '=' + value);
-        } else {
-          compilerOptionsArray.push('--' + option);
-        }
+        compilerFlags.push('--' + option);
       }
-    });
-
-    if (options.typeCheckOnly) {
-      return gulp
-        .src(srcs, {base: '.'})
-        .pipe(sourcemaps.init())
-        .pipe(preClosureBabel())
-        .on('error', (err) => handlePreClosureError(err, outputFilename))
-        .pipe(gulpClosureCompile(compilerOptionsArray))
-        .on('error', (err) => handleTypeCheckError(err))
-        .pipe(nop())
-        .on('end', resolve);
-    } else {
-      timeInfo.startTime = Date.now();
-      return gulp
-        .src(srcs, {base: '.'})
-        .pipe(sourcemaps.init())
-        .pipe(preClosureBabel())
-        .on('error', (err) =>
-          handlePreClosureError(err, outputFilename, options, resolve)
-        )
-        .pipe(gulpClosureCompile(compilerOptionsArray))
-        .on('error', (err) =>
-          handleCompilerError(err, outputFilename, options, resolve)
-        )
-        .pipe(rename(`${outputDir}/${outputFilename}`))
-        .pipe(
-          gulpIf(
-            !argv.pseudo_names && !options.skipUnknownDepsCheck,
-            checkForUnknownDeps()
-          )
-        )
-        .on('error', reject)
-        .pipe(postClosureBabel())
-        .pipe(sanitize())
-        .pipe(writeSourcemaps(options))
-        .pipe(gulp.dest('.'))
-        .on('end', resolve);
     }
   });
+  const jsFlags = globby.sync(transformedSrcFiles).map((src) => `--js=${src}`);
+  const entryPointFlags = entryModuleFilenames
+    .map((entryFile) =>
+      transformedSrcFiles.find((srcFile) => srcFile.endsWith(entryFile))
+    )
+    .map((entryPoint) => `--entry_point=${entryPoint}`);
+  const flags = compilerFlags.concat(entryPointFlags, jsFlags);
+  if (!options.typeCheckOnly) {
+    const outputFileFlag = `--js_output_file=${destFile}`;
+    const sourcemapFlag = `--create_source_map=${sourcemapFile}`;
+    flags.push(outputFileFlag, sourcemapFlag);
+  }
+  return flags;
+}
+
+/**
+ * @param {string[]|string} entryModuleFilenames
+ * @param {string} outputDir
+ * @param {string} outputFilename
+ * @param {!OptionsDef} options
+ * @param {{startTime?: number}} timeInfo
+ * @return {Promise<void>}
+ */
+async function compile(
+  entryModuleFilenames,
+  outputDir,
+  outputFilename,
+  options,
+  timeInfo
+) {
+  if (timeInfo) {
+    timeInfo.startTime = Date.now();
+  }
+  if (!(entryModuleFilenames instanceof Array)) {
```

Comment:
super nit: 
```suggestion
  if (!Array.isArray(entryModuleFilenames)) {
```

note: the only functional difference between the two array checks is that instanceof doesn't work through iframes, which clearly isn't relevant here. but it is still good practice to prefer!

Reply:
Done here and on line 371 as well.